include/camp/resource/hip.hpp

```diff
@@ -75,34 +78,57 @@ namespace resources
 
     class Hip
     {
-      static hipStream_t get_a_stream(int num)
+      static int get_current_device()
       {
-        static hipStream_t streams[16] = {};
-        static int previous = 0;
+        int dev = -1;
+        campHipErrchk(hipGetDevice(&dev));
+        return dev;
+      }
+
+      static int get_device_from_stream(hipStream_t stream)
+      {
+        return hipGetStreamDeviceId(stream);
+      }
 
-        static std::once_flag m_onceFlag;
-        static std::mutex m_mtx;
+      static hipStream_t get_a_stream(int num, int dev)
+      {
+        static constexpr int num_streams = 16;
+        struct Streams {
+          hipStream_t streams[num_streams] = {};
+          int previous = 0;
+
+          std::once_flag onceFlag;
+          std::mutex mtx;
+        };
+
+        static std::vector<Streams> devices([] {
+          int count = -1;
+          campHipErrchk(hipGetDeviceCount(&count));
+          return count;
+        }());
 
-        std::call_once(m_onceFlag, [] {
-          if (streams[0] == nullptr) {
-            for (auto &s : streams) {
+        std::call_once(devices[dev].onceFlag, [=] {
+          auto d{device_guard(dev)};
+          if (devices[dev].streams[0] == nullptr) {
+            for (auto &s : devices[dev].streams) {
               campHipErrchk(hipStreamCreate(&s));
             }
           }
         });
 
         if (num < 0) {
-          m_mtx.lock();
-          previous = (previous + 1) % 16;
-          m_mtx.unlock();
-          return streams[previous];
+          std::lock_guard<std::mutex> guard(devices[dev].mtx);
+          devices[dev].previous = (devices[dev].previous + 1) % num_streams;
+          num = devices[dev].previous;
+        } else {
+          num = num % num_streams;
         }
 
-        return streams[num % 16];
+        return devices[dev].streams[num];
       }
 
       // Private from-stream constructor
-      Hip(hipStream_t s, int dev = 0) : stream(s), device(dev) {}
+      Hip(hipStream_t s, int dev) : stream(s), device(dev) {}
```

Comment:
I see you wired in the way to get the device from the stream probably use that here as well?

Reply:
That makes some sense, but take a look at what I say about constructing a resource without any runtime calls.

Reply:
I think you're saying here that it makes more sense to put the conditional and get_device_from_stream call here rather than in HipFromStream. I'll do that.